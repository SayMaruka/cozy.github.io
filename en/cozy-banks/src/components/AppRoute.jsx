import React from 'react'
import { IndexRoute, Route, Redirect } from 'react-router'
import App from 'components/App'
import { isWebApp } from 'cozy-device-helper'

import { TransactionsPageWithBackButton } from 'ducks/transactions'
import { CategoriesPage } from 'ducks/categories'
import {
  Settings,
  AccountSettings,
  AccountsSettings,
  GroupsSettings,
  GroupSettings,
  NewGroupSettings,
  Configuration,
  Debug
} from 'ducks/settings'
import { Balance, BalanceDetailsPage } from 'ducks/balance'
import {
  DebugRecurrencePage,
  RecurrencesPage,
  RecurrencePage
} from 'ducks/recurrence'
import { TransferPage } from 'ducks/transfers'
import UserActionRequired from 'components/UserActionRequired'
import scrollToTopOnMount from 'components/scrollToTopOnMount'

// Use a function to delay instantation and have access to AppRoute.renderExtraRoutes
const AppRoute = () => (
  <Route component={UserActionRequired}>
    <Route component={App}>
      {isWebApp() && <Redirect from="/" to="balances" />}
      <Route path="balances">
        <IndexRoute component={scrollToTopOnMount(Balance)} />
        <Route
          path="details"
          component={scrollToTopOnMount(BalanceDetailsPage)}
        />
      </Route>
      <Route path="categories">
        <IndexRoute component={scrollToTopOnMount(CategoriesPage)} />
        <Route
          path=":categoryName/:subcategoryName"
          component={scrollToTopOnMount(TransactionsPageWithBackButton)}
        />
        <Route
          path=":categoryName"
          component={scrollToTopOnMount(CategoriesPage)}
        />
      </Route>
      <Route path="recurrence">
        <IndexRoute component={scrollToTopOnMount(RecurrencesPage)} />
        <Route
          path=":bundleId"
          component={scrollToTopOnMount(RecurrencePage)}
        />
      </Route>
      <Route path="settings">
        <Route
          path="groups/new"
          component={scrollToTopOnMount(NewGroupSettings)}
        />
        <Route
          path="groups/:groupId"
          component={scrollToTopOnMount(GroupSettings)}
        />
        <Route
          path="accounts/:accountId"
          component={scrollToTopOnMount(AccountSettings)}
        />
        <Route component={scrollToTopOnMount(Settings)}>
          <IndexRoute component={Configuration} />
          <Route path="accounts" component={AccountsSettings} />
          <Route path="groups" component={GroupsSettings} />
          <Route path="configuration" component={Configuration} />
          <Route path="debug" component={Debug} />
        </Route>
      </Route>
      <Route path="transfers" component={scrollToTopOnMount(TransferPage)} />
      <Route path="transfers/:slideName" component={TransferPage} />
      <Route
        path="recurrencedebug"
        component={scrollToTopOnMount(DebugRecurrencePage)}
      />
      {AppRoute.renderExtraRoutes()}
      {isWebApp() && <Redirect from="*" to="balances" />}
    </Route>
  </Route>
)

// Ability to overrides easily
AppRoute.renderExtraRoutes = () => null

export default AppRoute
