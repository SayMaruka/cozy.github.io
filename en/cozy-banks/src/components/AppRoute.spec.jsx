import AppRoute from './AppRoute'
import React from 'react'
import { Route } from 'react-router'

const NewComponent = () => <div>New component</div>

describe('App route', () => {
  afterEach(() => {
    jest.restoreAllMocks()
  })

  it('should be a function returning a route', () => {
    const route = AppRoute()
    expect(route).toMatchSnapshot()
  })

  it('should have renderExtraRoutes', () => {
    jest
      .spyOn(AppRoute, 'renderExtraRoutes')
      .mockReturnValue(<Route component={NewComponent} path="extra-route" />)
    const route = AppRoute()
    expect(route).toMatchSnapshot()
  })
})
