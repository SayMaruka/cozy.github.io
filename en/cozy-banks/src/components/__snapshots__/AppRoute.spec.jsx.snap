// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App route should be a function returning a route 1`] = `
<Route
  component={[Function]}
>
  <Route
    component={[Function]}
  >
    <Redirect
      from="/"
      to="balances"
    />
    <Route
      path="balances"
    >
      <IndexRoute
        component={[Function]}
      />
      <Route
        component={[Function]}
        path="details"
      />
    </Route>
    <Route
      path="categories"
    >
      <IndexRoute
        component={[Function]}
      />
      <Route
        component={[Function]}
        path=":categoryName/:subcategoryName"
      />
      <Route
        component={[Function]}
        path=":categoryName"
      />
    </Route>
    <Route
      path="recurrence"
    >
      <IndexRoute
        component={[Function]}
      />
      <Route
        component={[Function]}
        path=":bundleId"
      />
    </Route>
    <Route
      path="settings"
    >
      <Route
        component={[Function]}
        path="groups/new"
      />
      <Route
        component={[Function]}
        path="groups/:groupId"
      />
      <Route
        component={[Function]}
        path="accounts/:accountId"
      />
      <Route
        component={[Function]}
      >
        <IndexRoute
          component={[Function]}
        />
        <Route
          component={[Function]}
          path="accounts"
        />
        <Route
          component={[Function]}
          path="groups"
        />
        <Route
          component={[Function]}
          path="configuration"
        />
        <Route
          component={[Function]}
          path="debug"
        />
      </Route>
    </Route>
    <Route
      component={[Function]}
      path="transfers"
    />
    <Route
      component={[Function]}
      path="transfers/:slideName"
    />
    <Route
      component={[Function]}
      path="recurrencedebug"
    />
    <Redirect
      from="*"
      to="balances"
    />
  </Route>
</Route>
`;

exports[`App route should have renderExtraRoutes 1`] = `
<Route
  component={[Function]}
>
  <Route
    component={[Function]}
  >
    <Redirect
      from="/"
      to="balances"
    />
    <Route
      path="balances"
    >
      <IndexRoute
        component={[Function]}
      />
      <Route
        component={[Function]}
        path="details"
      />
    </Route>
    <Route
      path="categories"
    >
      <IndexRoute
        component={[Function]}
      />
      <Route
        component={[Function]}
        path=":categoryName/:subcategoryName"
      />
      <Route
        component={[Function]}
        path=":categoryName"
      />
    </Route>
    <Route
      path="recurrence"
    >
      <IndexRoute
        component={[Function]}
      />
      <Route
        component={[Function]}
        path=":bundleId"
      />
    </Route>
    <Route
      path="settings"
    >
      <Route
        component={[Function]}
        path="groups/new"
      />
      <Route
        component={[Function]}
        path="groups/:groupId"
      />
      <Route
        component={[Function]}
        path="accounts/:accountId"
      />
      <Route
        component={[Function]}
      >
        <IndexRoute
          component={[Function]}
        />
        <Route
          component={[Function]}
          path="accounts"
        />
        <Route
          component={[Function]}
          path="groups"
        />
        <Route
          component={[Function]}
          path="configuration"
        />
        <Route
          component={[Function]}
          path="debug"
        />
      </Route>
    </Route>
    <Route
      component={[Function]}
      path="transfers"
    />
    <Route
      component={[Function]}
      path="transfers/:slideName"
    />
    <Route
      component={[Function]}
      path="recurrencedebug"
    />
    <Route
      component={[Function]}
      path="extra-route"
    />
    <Redirect
      from="*"
      to="balances"
    />
  </Route>
</Route>
`;
